### PR TITLE
MIR drive-by cleanups

### PR DESCRIPTION
**Pull request by WaffleLapkin**
Some random drive-by cleanups I did while working with MIR/THIR.

### CODE REVIEW

File: compiler/rustc_mir_build/src/thir/cx/expr.rs [hunk 1]
```diff
@@ -130,6 +130,7 @@ impl<'tcx> Cx<'tcx> {
                 ExprKind::Pointer { cast: PointerCast::Unsize, source: self.thir.exprs.push(expr) }
             }
             Adjust::Pointer(cast) => ExprKind::Pointer { cast, source: self.thir.exprs.push(expr) },
+            Adjust::NeverToAny if adjustment.target.is_never() => return expr,
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/thir/cx/expr.rs:133]: is this testable in a mir-opt test without `#[custom_mir]`?

File: compiler/rustc_middle/src/mir/syntax.rs [hunk 2]
```diff
@@ -749,6 +749,29 @@ pub enum TerminatorKind<'tcx> {
     },
 }
 
+impl TerminatorKind<'_> {
```
Comment (JakobDegen) [inline compiler/rustc_middle/src/mir/syntax.rs:752]: Can you put this in a different file? This one's mostly for type definitions

File: compiler/rustc_mir_build/src/thir/cx/expr.rs (hunk 1 above)
Comment (WaffleLapkin) [reply compiler/rustc_mir_build/src/thir/cx/expr.rs:133]: It looks like it is: https://github.com/rust-lang/rust/pull/111501#issuecomment-1545543420 😅

File: compiler/rustc_middle/src/mir/syntax.rs (hunk 2 above)
Comment (WaffleLapkin) [reply compiler/rustc_middle/src/mir/syntax.rs:752]: There are similar functions in this file though, like `MirPhase::name`...

Not sure where to best put this

Comment (JakobDegen) [reply compiler/rustc_middle/src/mir/syntax.rs:752]: Hmm, ok. We should probably move those out too, don't worry about it for this PR though

**Comment by oli-obk** [comment]
r=me with CI happy

File: compiler/rustc_middle/src/mir/syntax.rs (hunk 2 above)
Comment (cjgillot) [reply compiler/rustc_middle/src/mir/syntax.rs:752]: The mir::terminator file looks like a proper place.

**Comment by WaffleLapkin** [comment]
So the removal of `!` -> `!` coercion changes the place where `!` is used, changing some miri diagnostics (namely "reached unreachable code"). I _think_ it's fine, but just in case cc @RalfJung. I've blessed this test: https://github.com/rust-lang/rust/pull/111501/commits/693d36fc009a2032dd1605aef2d7bb854f8e15fc#diff-483e7d9f8800c2c4dab29c8c5504034bce3e57ae01bc087aefc181c0e93f457a

File: src/tools/miri/tests/fail/never_say_never.rs [hunk 3]
```diff
@@ -6,10 +6,8 @@
 
 fn main() {
     let y = &5;
-    let x: ! = unsafe {
-        *(y as *const _ as *const !) //~ ERROR: entering unreachable code
-    };
-    f(x)
+    let x: ! = unsafe { *(y as *const _ as *const !) };
+    f(x) //~ ERROR: entering unreachable code
 }
```
Comment (oli-obk) [inline src/tools/miri/tests/fail/never_say_never.rs:11]: This is surprising. How does the MIR change here? The previous MIR didn't even contain an `f` call, even with `-Zmir-opt-level=0`

Comment (WaffleLapkin) [reply src/tools/miri/tests/fail/never_say_never.rs:11]: The change is essentially
1. `f` now has `return`, not `unreachable`
2. `main` now has additional `_3: !`
3. `main` now has `_3 = f(const ZeroSized: !);`, not `unreachable;`

Comment (WaffleLapkin) [reply src/tools/miri/tests/fail/never_say_never.rs:11]: Hm, it looks like the difference is already present after building MIR:

```diff
--- ./nightly_mir_dump/t.f.001-000.built.after.mir      2023-05-15 14:25:01.537195340 +0000
+++ ./b_mir_dump/t.f.001-000.built.after.mir    2023-05-15 14:25:07.469143513 +0000
@@ -3,23 +3,9 @@
 fn f(_1: !) -> ! {
     debug x => _1;                       // in scope 0 at ./t.rs:9:6: 9:7
     let mut _0: !;                       // return place in scope 0 at ./t.rs:9:15: 9:16
-    let mut _2: !;                       // in scope 0 at ./t.rs:9:17: 11:2
-    let mut _3: !;                       // in scope 0 at ./t.rs:10:5: 10:6
 
     bb0: {
-        StorageLive(_2);                 // scope 0 at ./t.rs:9:17: 11:2
-        StorageLive(_3);                 // scope 0 at ./t.rs:10:5: 10:6
-        _3 = _1;                         // scope 0 at ./t.rs:10:5: 10:6
-        unreachable;                     // scope 0 at ./t.rs:10:5: 10:6
-    }
-
-    bb1: {
-        StorageDead(_3);                 // scope 0 at ./t.rs:10:5: 10:6
-        unreachable;                     // scope 0 at ./t.rs:9:17: 11:2
-    }
-
-    bb2: {
-        StorageDead(_2);                 // scope 0 at ./t.rs:11:1: 11:2
+        _0 = _1;                         // scope 0 at ./t.rs:10:5: 10:6
         return;                          // scope 0 at ./t.rs:11:2: 11:2
     }
 }
```

Comment (RalfJung) [reply src/tools/miri/tests/fail/never_say_never.rs:11]: I don't understand why `NeverToAny` should get a special case for `! -> !`, but I also don't mind the change in diagnostics here.

Comment (RalfJung) [reply src/tools/miri/tests/fail/never_say_never.rs:11]: That's somewhat surprising but well, I am not familiar with those parts of the compiler at all.^^
Still, @rust-lang/types should probably take a look.

**Pull request merged**